ratis-common/src/main/java/org/apache/ratis/util/LifeCycle.java

```diff
@@ -100,8 +100,8 @@ static void put(State key, Map<State, List<State>> map, State... values) {
       put(PAUSING,   predecessors, RUNNING);
       put(PAUSED,    predecessors, PAUSING);
       put(EXCEPTION, predecessors, STARTING, PAUSING, RUNNING);
-      put(CLOSING,   predecessors, STARTING, RUNNING, PAUSING, PAUSED, EXCEPTION);
-      put(CLOSED,    predecessors, NEW, CLOSING);
+      put(CLOSING,   predecessors, NEW, STARTING, RUNNING, PAUSING, PAUSED, EXCEPTION);
+      put(CLOSED,    predecessors, CLOSING);
```

Comment:
@ChenSammi , as stated in the javadoc, NEW means that the machine is newly created and holds zero resource so that it can be transit to CLOSED immediately.

BTW, we should not change LifeCycle since it is used in many places.

Reply:
@ChenSammi , which class?  We should change the state to STARTING before acquiring the storage LOCK.

Reply:
The RaftStorageImpl LOCK.

Reply:
That's is a good point.  RaftStorageImpl should not grab the LOCK in the constructor.  Let me see how to fix it.

Reply:
Filed https://issues.apache.org/jira/browse/RATIS-1609

Reply:
Another one is NettyRpcService. It starts the netty server in constructor not in the startImpl call.  So for the RaftServer using nettyRpcService, even this RaftServer is in NEW state, we should call it's closeImpl.  But currently it's not called on NEW state.  I didn't find a better solution for this one. So I basically suggest change the checkStateAndClose implemention as following. What do you think? 

```
if (compareAndTransition(State.NEW, State.CLOSED)) {
      return closeMethod.get();
 }
```

Reply:
We should fix NettyRpcService.

Reply:
We should fix NettyRpcService.

Reply:
Filed https://issues.apache.org/jira/browse/RATIS-1610

Reply:
There is one hidden issue if we move the NettyRpcService to startImpl,  getInetSocketAddress is called before startImpl.  getInetSocketAddress will try to get the server port.  The default configured port is 0. So if server is not started, we cannot know the actual port of this server.